File: etl/scripts/close_bet_config.py
"""종가베팅 전략값 config 로더 — 웹(/admin/settings)이 쓰고 배치가 읽는 단일 소스.

파일: scripts/close_bet.json (broker-web/lib/close-bet-config.ts 와 같은 파일).
배치(run_close_bet / run_close_bet_exit / report_close_bet_order)가 실행 시점에 load()
로 읽어 전략값을 얻는다. 파일이 없거나 키가 빠지면 하드코딩 기본값(=기존 동작)으로 폴백해
config 없이도 배치가 그대로 돈다. 값이 범위를 벗어나면 ValueError 로 배치를 abort 해
깨진 값으로 실주문 하는 사고를 막는다(웹 저장단·py 로더 이중 방어).

사용 (scripts, _bootstrap 뒤):
    from close_bet_config import load
    cfg = load()
    threshold = cfg["score_threshold"]  # int
    budget = cfg["budget_by_count"][n]  # int, n∈{1,2,3}
    cap_max = cfg["cap_max"]            # int, 원. 시총 상한(미만)
    tv_min = cfg["turnover_min"]        # int, 원. 전일 거래대금 하한(이상)
"""
from __future__ import annotations

import json
from pathlib import Path

# 하드코딩 기본값 = 기존 배치 값(run_close_bet default 70, exit tp/sl, _BUDGET_BY_COUNT).
# config 파일/키가 없을 때 폴백 → config 도입 전과 동일 동작 보존.
DEFAULTS: dict = {
    "score_threshold": 70,
    "tp": 0.05,
    "sl": 0.03,
    "budget_by_count": {1: 3_000_000, 2: 2_000_000, 3: 5_000_000 // 3},
    # 여기 값은 자리표시자다. 실제 운영값은 close_bet.json(gitignore, 저장소가 공개라
    # 검증된 파라미터를 커밋하지 않는다) 에 있고 근거는 research/private/close_bet_overnight.
    # 파일이 없으면 이 값으로 폴백하는데 필터가 사실상 안 걸리므로, load() 가 경고를 낸다.
    "cap_max": 10_000_000_000_000,   # 10조 = 사실상 상한 없음
    "turnover_min": 1,               # 1원 = 사실상 하한 없음
    # 청산 워커(run_close_bet_exit)의 강제청산 시각. ps1 인자가 아니라 여기가 단일 소스라
    # 값만 바꾸면 다음 기동부터 반영된다.
    # ⚠ 백스톱 배치 시각(ops/scheduled-tasks/close-bet-force-exit.xml 의 StartBoundary)은
    #   스케줄러 소관이라 자동 연동되지 않는다. 이 값을 바꾸면 그 XML 도 같이 바꿀 것.
    "exit_time": "15:19:00",
}

_COUNTS = (1, 2, 3)


def PATH() -> Path:
    """config 파일 경로 (이 모듈과 같은 scripts/ 디렉터리)."""
    return Path(__file__).resolve().parent / "close_bet.json"


def _is_int(x) -> bool:
    return isinstance(x, int) and not isinstance(x, bool)


def _is_hms(x) -> bool:
    """'HH:MM:SS' 형식이고 실재하는 시각인지. 워커가 datetime.replace 로 쓰므로 범위까지 본다."""
    if not isinstance(x, str):
        return False
    parts = x.split(":")
    if len(parts) != 3 or not all(len(p) == 2 and p.isdigit() for p in parts):
        return False
    h, m, s = (int(p) for p in parts)
    return h <= 23 and m <= 59 and s <= 59


def _validate(cfg: dict) -> None:
    """범위 검증. 실패 시 ValueError(배치 abort)."""
    st = cfg["score_threshold"]
    if not _is_int(st) or not (0 <= st <= 100):
        raise ValueError(f"score_threshold must be int 0~100, got {st!r}")
    for k in ("tp", "sl"):
        v = cfg[k]
        if v is None:  # null = 장중 TP/SL 판정 끔(09:01 강제청산만)
            continue
        if isinstance(v, bool) or not isinstance(v, (int, float)) or not (0 <= v <= 1):
            raise ValueError(f"{k} must be number 0~1 or null, got {v!r}")
    for k in ("cap_max", "turnover_min"):
        v = cfg[k]
        if not _is_int(v) or v <= 0:
            raise ValueError(f"{k} must be positive int (원), got {v!r}")
    if not _is_hms(cfg["exit_time"]):
        raise ValueError(f"exit_time must be 'HH:MM:SS', got {cfg['exit_time']!r}")
    budget = cfg["budget_by_count"]
    if not isinstance(budget, dict):
        raise ValueError(f"budget_by_count must be dict, got {budget!r}")
    for n in _COUNTS:
        if n not in budget:
            raise ValueError(f"budget_by_count missing key {n}")
        amt = budget[n]
        if not _is_int(amt) or amt <= 0:
            raise ValueError(f"budget_by_count[{n}] must be positive int, got {amt!r}")


def load(path: Path | None = None) -> dict:
    """config 로드 → 범위검증된 dict. 파일/키 없으면 기본값 폴백.

    close_bet.json 은 gitignore 라 새 체크아웃에는 없다. 그때 폴백되는 DEFAULTS 의
    cap_max·turnover_min 은 자리표시자(사실상 필터 없음)라 그대로 매매하면 안 된다.
    조용히 넘어가지 않도록 경고를 낸다 — close_bet.example.json 을 복사해 값을 채울 것.
    """
    p = path or PATH()
    raw: dict = {}
    if p.exists():
        raw = json.loads(p.read_text(encoding="utf-8"))
    else:
        print(f"[close_bet] ⚠️ config 없음({p}) — 기본값 폴백. "
              f"시총·거래대금 필터가 사실상 비활성이다. close_bet.example.json 참고")

    cfg = {
        "score_threshold": raw.get("score_threshold", DEFAULTS["score_threshold"]),
        "tp": raw.get("tp", DEFAULTS["tp"]),
        "sl": raw.get("sl", DEFAULTS["sl"]),
        "cap_max": raw.get("cap_max", DEFAULTS["cap_max"]),
        "turnover_min": raw.get("turnover_min", DEFAULTS["turnover_min"]),
        "exit_time": raw.get("exit_time", DEFAULTS["exit_time"]),
    }
    if "budget_by_count" in raw:
        # JSON 키는 문자열 → int 키로 변환(배치는 int n 으로 조회).
        cfg["budget_by_count"] = {int(k): v for k, v in raw["budget_by_count"].items()}
    else:
        cfg["budget_by_count"] = dict(DEFAULTS["budget_by_count"])

    _validate(cfg)
    return cfg


if __name__ == "__main__":  # 셀프체크
    import tempfile

    def _tmp(cfg: dict) -> Path:
        p = Path(tempfile.mktemp(suffix=".json"))
        p.write_text(json.dumps(cfg), encoding="utf-8")
        return p

    # 정상
    ok = load(_tmp({"score_threshold": 50, "tp": 0.05, "sl": 0.03,
                    "cap_max": 100_000_000_000, "turnover_min": 2_000_000_000,
                    "exit_time": "09:01:00",
                    "budget_by_count": {"1": 3000000, "2": 2000000, "3": 1666666}}))
    assert ok["score_threshold"] == 50 and ok["budget_by_count"][1] == 3000000
    assert ok["cap_max"] == 100_000_000_000 and ok["turnover_min"] == 2_000_000_000
    assert ok["exit_time"] == "09:01:00"

    # 파일 없음 → 기본값
    assert load(Path(tempfile.gettempdir()) / "nope.json") == DEFAULTS

    # 범위 초과
    for bad in ({"score_threshold": 101}, {"tp": 2}, {"sl": -1}, {"cap_max": 0},
                {"turnover_min": -1}, {"exit_time": "9:1"}, {"exit_time": "25:00:00"},
                {"budget_by_count": {"1": 1, "2": 0, "3": 1}}):
        try:
            load(_tmp(bad))
            raise AssertionError(f"expected ValueError for {bad}")
        except ValueError:
            pass

    print("[close_bet_config] self-check OK")

File: broker-web/lib/close-bet-config.ts
import path from "node:path";
import fs from "node:fs/promises";

// 종가베팅 전략값 config. 배치(etl/scripts/close_bet_config.py)와 같은 파일을 읽고 쓴다.
// 저장 포맷은 소수(tp/sl 0~1) — UI 는 %로 표시/입력하고 폼 계층에서 ÷100 변환해 여기로 넘긴다.
// 범위검증은 py 로더(close_bet_config._validate)와 동일 도메인. 웹 저장단·py 로더 이중 방어.
// ponytail: 로컬 단일사용자 도구라 파일 락 없음.

export type BudgetByCount = Record<"1" | "2" | "3", number>;
export type CloseBetConfig = {
  score_threshold: number;
  tp: number | null; // null = 장중 TP/SL 판정 끔(익일 강제청산만)
  sl: number | null;
  cap_max: number; // 원. 시총 상한(미만)
  turnover_min: number; // 원. 전일 거래대금 하한(이상)
  cap_min_pct: number; // 0~1(1 미만). 전일 전 종목 시총 하위 이 비율은 15시 스코어링에서 제외. 0 = 끔
  exit_time: string; // "HH:MM:SS". 청산 워커의 강제청산 시각(단일 소스)
  budget_by_count: BudgetByCount;
};

// 하드코딩 기본값 = py 로더 DEFAULTS 와 동일(파일 없을 때 폴백).
export const DEFAULTS: CloseBetConfig = {
  score_threshold: 70,
  tp: 0.05,
  sl: 0.03,
  // 자리표시자. 실제 운영값은 close_bet.json(gitignore) — py 로더 DEFAULTS 와 같은 규칙.
  cap_max: 10_000_000_000_000,
  turnover_min: 1,
  cap_min_pct: 0,
  exit_time: "15:19:00",
  budget_by_count: { "1": 3_000_000, "2": 2_000_000, "3": Math.floor(5_000_000 / 3) },
};

const COUNTS: (keyof BudgetByCount)[] = ["1", "2", "3"];

function repoRoot() {
  return path.join(process.cwd(), "..");
}
export const PATH = () => path.join(repoRoot(), "etl", "scripts", "close_bet.json");

// "HH:MM:SS" 이고 실재하는 시각인지 (py 로더 _is_hms 와 같은 도메인).
export function isHms(v: unknown): boolean {
  if (typeof v !== "string" || !/^\d{2}:\d{2}:\d{2}$/.test(v)) return false;
  const [h, m, s] = v.split(":").map(Number);
  return h <= 23 && m <= 59 && s <= 59;
}

export function validate(cfg: CloseBetConfig): void {
  const st = cfg.score_threshold;
  if (!Number.isInteger(st) || st < 0 || st > 100) {
    throw new Error("매수 기준점수는 0~100 사이 정수여야 합니다");
  }
  for (const [k, label] of [["tp", "익절"], ["sl", "손절"]] as const) {
    const v = cfg[k];
    if (v === null) continue; // 빈칸 = 사용 안 함
    if (typeof v !== "number" || !Number.isFinite(v) || v < 0 || v > 1) {
      throw new Error(`${label}은 0~100% 사이 값이거나 비어 있어야 합니다`);
    }
  }
  for (const [k, label] of [["cap_max", "시가총액 상한"], ["turnover_min", "거래대금 하한"]] as const) {
    const v = cfg[k];
    if (!Number.isInteger(v) || v <= 0) {
      throw new Error(`${label}은 0보다 큰 금액이어야 합니다`);
    }
  }
  const pct = cfg.cap_min_pct;
  if (typeof pct !== "number" || !Number.isFinite(pct) || pct < 0 || pct >= 1) {
    throw new Error("시가총액 하한(하위 %)은 0 이상 100 미만이어야 합니다");
  }
  if (!isHms(cfg.exit_time)) {
    throw new Error("청산 시각은 HH:MM:SS 형식이어야 합니다 (예: 09:01:00)");
  }
  const budget = cfg.budget_by_count;
  if (!budget || typeof budget !== "object") {
    throw new Error("종목당 예산 형식이 올바르지 않습니다");
  }
  for (const n of COUNTS) {
    const amt = budget[n];
    if (!Number.isInteger(amt) || amt <= 0) {
      throw new Error(`${n}종목일 때 예산은 0보다 큰 정수여야 합니다`);
    }
  }
}

// 파일 읽기 → 키 누락 시 기본값 폴백 → 검증. 파일 없으면 DEFAULTS.
export async function read(): Promise<CloseBetConfig> {
  let raw: Partial<CloseBetConfig> = {};
  try {
    raw = JSON.parse(await fs.readFile(PATH(), "utf-8"));
  } catch {
    return { ...DEFAULTS, budget_by_count: { ...DEFAULTS.budget_by_count } };
  }
  const cfg: CloseBetConfig = {
    score_threshold: raw.score_threshold ?? DEFAULTS.score_threshold,
    // null 은 "끔"이라 ?? 로 기본값을 덮으면 안 된다 — 키 존재 여부로 판단한다.
    tp: "tp" in raw ? (raw.tp as number | null) : DEFAULTS.tp,
    sl: "sl" in raw ? (raw.sl as number | null) : DEFAULTS.sl,
    cap_max: raw.cap_max ?? DEFAULTS.cap_max,
    turnover_min: raw.turnover_min ?? DEFAULTS.turnover_min,
    cap_min_pct: raw.cap_min_pct ?? DEFAULTS.cap_min_pct,
    exit_time: raw.exit_time ?? DEFAULTS.exit_time,
    budget_by_count: raw.budget_by_count
      ? (raw.budget_by_count as BudgetByCount)
      : { ...DEFAULTS.budget_by_count },
  };
  validate(cfg);
  return cfg;
}

// 검증 통과 시에만: 직전 버전 .bak 백업 후 저장. 검증 실패 시 파일 불변(throw).
export async function write(cfg: CloseBetConfig): Promise<void> {
  validate(cfg);
  const p = PATH();
  try {
    const prev = await fs.readFile(p, "utf-8");
    await fs.writeFile(p + ".bak", prev, "utf-8");
  } catch {
    // 원본 없으면 백업 생략(최초 저장).
  }
  await fs.writeFile(p, JSON.stringify(cfg, null, 2) + "\n", "utf-8");
}
